File: Letter Combinations of a Phone Number.cpp
class Solution {
public:
    string map[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> letterCombinations(string digits) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<string> res;
        res.push_back("");
        if (digits.size() == 0) return res;
        string chs;
        
        for (int i = 0; i < digits.length(); i++) {
            vector<string> tmp;
            chs = map[digits[i] - '0'];
            while (!res.empty()) {
                string s = res.back();
                res.pop_back();
                for (int i = 0; i < chs.length(); i++) {
                    tmp.push_back(string(s+chs[i]));
                }       
            }
            res = tmp;
        }
        return res;
    }
};

File: Letter Combinations of a Phone Number.cs
public class Solution {
    private static readonly string[] map
        = new string[] {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    
    public IList<string> LetterCombinations(string digits) {
        var result = new List<string>(); 
        if (string.IsNullOrEmpty(digits)) {
            return result;
        }
        
        result.Add("");
        foreach (var ch in digits) {
            var list = new List<string>();
            var letters = map[ch - '0'];
            foreach (var ans in result) {
                foreach (var letter in letters) {
                    list.Add(ans + letter);
                }
            }
            result = list;
        }
        return result;
    }
}
